### PR TITLE
Starting on webpki CRL parsing support.

### PR DESCRIPTION
**Pull request by cpu**
This branch adds a new `crl.rs` module that offers initial bare bones support for parsing RFC 5280 web PKI profile certificate revocation lists. No support is added for processing CRLs for revocation decisions, this is simply the code required to parse enough DER encoding that we'll have the pieces we need to build additional revocation checking support.

The parser code is modelled after the approach used for certificates, and with the `x509.rs` refactoring included in this branch, can share some common code for processing extensions.

Initially I've taken a fairly conservative approach and have excluded support for:

* CRLs with versions other than 2.
* CRLs missing a next update field.
* CRLs missing extensions required by the web PKI profile.
* "Delta" CRLs where only new entries compared to a previous CRL are included.
* "Indirect" CRLs where serials issued by a CA other than the one that signed the CRL are included.
* CRLs with revoked certificates encoded with a negative serial number.

For the serial number handling, we'll probably want to relax this to match the handling of trust anchor/EE serials in webpki but it will require some care to make sure we perform CRL lookups correctly if we've "fixed up" negative serials with the incorrect MSB.

~Test coverage so far is strictly based on hand crafted CRLs made with [der-ascii](https://github.com/google/der-ascii). I haven't yet looked at how this parser fares with a corpus of "real world" CRLs. I suspect doing so will uncover that we'll need to support some of the "features" I've omitted so far.~ I've since tested this code with a corpus of 1,541 CRLs fetched from CCADB in addition to the unit test coverage created with der-ascii.

To move forward towards real revocation checking we'll also need to think about how best to load a number of CRLs into a webpki validation context, and how to use our trust roots to verify the CRL signed data. 

Related to https://github.com/rustls/rustls/issues/1164

Resolves https://github.com/rustls/webpki/issues/54

### CODE REVIEW

**Comment by cpu** [comment]
> ci / package (pull_request) Failing after 12s

Looks like I need to add my new `.rs` file to the Cargo manifest `include` to fix this failure.

**Comment by cpu** [comment]
> cpu force-pushed the cpu-crl-support-pt1 branch from 962a1ee to 7055b62 

To hopefully make this easier to review I split up the work adding the CRL parsing into 4 sub-commits: 

1. first adding the bare essentials to parse the outer structure, 
2. then adding support for the CRL list extensions, 
3. then adding support to parse and iterate the revoked entries, 
4. and then finally adding support to parse the revoked entry extensions.

> Looks like I need to add my new .rs file to the Cargo manifest include to fix this failure.

Handled in c462e5c537b38e4ed4164e89315b408b2cfb58cb and b1171a3c6b0ad0bfbd89e042c969c42b20cbbcb0. CI is all green now :tada:

> Test coverage so far is strictly based on hand crafted CRLs made with [der-ascii](https://github.com/google/der-ascii). I haven't yet looked at how this parser fares with a corpus of "real world" CRLs. I suspect doing so will uncover that we'll need to support some of the "features" I've omitted so far.

I went on a brief side-quest and wrote tooling to parse CSV reports from [CCADB](https://www.ccadb.org/) to find & download all complete and partitioned CRLs for the trusted roots and disclosed intermediates in the Mozilla root program. This produced a decent sized corpus of CRL data to experiment with. I'll be looking at that in the coming days with the goal of flipping this PR to "ready to review" shortly after. I will also publish the tooling I wrote in case its helpful for others.

**Comment by cpu** [comment]
> I went on a brief side-quest and wrote tooling to parse CSV reports from [CCADB](https://www.ccadb.org/) to find & download all complete and partitioned CRLs for the trusted roots and disclosed intermediates in the Mozilla root program. This produced a decent sized corpus of CRL data to experiment with. I'll be looking at that in the coming days with the goal of flipping this PR to "ready to review" shortly after. I will also publish the tooling I wrote in case its helpful for others.

This was a useful exercise and it shook out two problems with the code in this branch. One I've fixed, and one that will require a bit more thought.

The first issue was parsing revoked certificate entries that encoded the `Extensions OPTIONAL` field as a `SEQUENCE` of length 0 instead of omitting the sequence. Based on my understanding of the ITU X.690 BER/DER encoding rules this is a mis-encoding,  however, it was prevalent in the CRL dataset I was processing (at least ~20 occurrences) and I found other CRL parsers (e.g. Go's) were tolerant of this choice. I've updated the code in this branch to allow this with a comment indicating why.

The second issue was discovering that Ring's `read_tag_and_get_value` function from `io/der.rs` only supported values with lengths that could be encoded in [at most two bytes](https://github.com/briansmith/ring/blob/95948b3977013aed16db92ae32e6b8384496a740/src/io/der.rs#L79-L100). I couldn't find a definitive reason in the development history but suspect this was a cautious choice that worked for the domain of smaller certificate data Ring was operating with. Unfortunately, CRLs can be quite large (up to 50mb in my corpus) and thus require processing DER values with lengths that may span three or four bytes. To unblock my testing I made [a crude patch](https://github.com/cpu/ring/commit/ae80a1abbef3b0dea0c48fa20b51b515e995ae32) to ring and found no further roadblocks with this branch. I don't think this patch is an appropriate change as-is, we likely want to restrict these large values to the domain of CRL processing and not allow large values in other domains.

With the two changes mentioned above in place I was able to parse all 1,541 CRLs I downloaded from CCADB without error. Upon further reflection I'm not entirely sure this dataset will be representative of CRLs used for **client certificates** but I think it was a good place to start since it was an easily accessible corpus.

I think this PR is probably ready for review now and I'll be thinking about next steps concurrent to addressing any review feedback.

File: src/crl.rs
Comment (djc) [inline src/crl.rs:204]: Maybe `revoked_certs` could be an empty `untrusted::Input` instead of wrapping it in `Option` in the case it ends up being absent?

File: src/crl.rs
Comment (djc) [inline src/crl.rs:69]: Given the comment for `Unspecified`, should we make this non-`Option`al instead?

File: src/crl.rs
Comment (djc) [inline src/crl.rs:101]: This type contains nothing specific to revoked certificates, maybe it should have a less specific name?

File: src/crl.rs
Comment (djc) [inline src/crl.rs:110]: This docstring seems to want to evoke brackets to point at items, but I think rustdoc requires the use of backticks (like ```[`RevokedCert`]```?). Also it would be nice to use them consistently, including for the argument name `f`.

File: src/crl.rs
Comment (djc) [inline src/crl.rs:115]: FWIW I think this pattern is extremely obtuse but I haven't dug in enough to figure out if it can be replaced by something more idiomatic without losing some key guarantee.

File: src/error.rs
Comment (djc) [inline src/error.rs:140]: BTW, for both `InvalidSerialNumber` and `InvalidCrlNumber` it would be nice to stick it in some kind of alphabetic order as much as possible (or maybe include a commit to reorder up front/in a separate PR?).

File: src/crl.rs
Comment (djc) [inline src/crl.rs:425]: Style: import the trait, please?

File: tests/crl_tests.rs
Comment (djc) [inline tests/crl_tests.rs:19]: Please squash this into the commit that introduced the formatting issue.

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: This fell away when I reworked the iteration.

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: That's fair. I think in this case at least it's easy enough to do what we're looking for with `IntoIterator` and a helper. I've reworked using that approach.

File: src/x509.rs
Comment (cpu) [reply src/x509.rs]: Done :+1:

File: src/cert.rs
Comment (cpu) [reply src/cert.rs:147]: I was able to make this change. It seems viable to me :+1: Thanks.

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: Done.

File: src/der.rs
Comment (cpu) [reply src/der.rs]: Switched to in-line use of `ring::io::der::positive_integer`.

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: This fell away when I reworked with the iterator, but I made sure the new `find_serial` helper has consistent rustdoc linking.

File: src/error.rs
Comment (cpu) [reply src/error.rs]: I reordered up-front in a separate commit before adding any new entries. It felt reasonable to do that as part of this PR.

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: Done.

File: tests/crl_tests.rs
Comment (cpu) [reply tests/crl_tests.rs:21]: Done.

File: src/crl.rs
Comment (djc) [reply src/crl.rs:69]: That sounds reasonable. Maybe add a comment to that effect?

File: src/cert.rs
Comment (djc) [inline src/cert.rs]: Style nit: IMO a `match` for this is slightly easier to read and more concise.

File: src/x509.rs
Comment (djc) [inline src/x509.rs:62]: Style nit: `match`, as elsewhere. Maybe this should be a `Extension::unsupported() -> Result<(), Error>` method? I think that method would have 4 potential call sites by the end of this PR.

File: src/crl.rs
Comment (djc) [inline src/crl.rs:107]: Style suggestion: this is the kind of thing I'd write as a match. Or, consider `(!self.reader.at_end()).then_some(|| parse_revoked_cert(&mut self.reader))`? Ahh, I guess our MSRV doesn't support that.

File: src/crl.rs
Comment (djc) [inline src/crl.rs:358]: Style nit: in order to avoid rightward drift, I'd probably invert the condition here to return early (and maybe do the same for the `ext_seq.is_empty()` check)?

**Review by djc — APPROVED** [review]
Mostly LGTM, but I do think @ctz should look at this for a bit before we merge, too.

File: src/crl.rs
Comment (djc) [reply src/crl.rs:107]: (Submitted #62 to pull up the MSRV.)

File: src/cert.rs
Comment (cpu) [reply src/cert.rs]: Switched to a `match` in the impl of `Extension::unsupported()`.

File: src/x509.rs
Comment (cpu) [reply src/x509.rs]: Good idea, thanks! I implemented the `Extension::unsupported()` idea and threaded its use through subsequent commits.

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: That's much nicer :sparkles: Implemented in my recent update.

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: Done :+1:

**Review by ctz — APPROVED** [review]
LGTM!

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: Yup!

**Comment by cpu** [comment]
Thanks for the reviews folks :rocket: 

I'm going to merge this but there's lots of opportunity to iterate further in the subsequent PRs if anyone has additional feedback on these parts.

File: src/crl.rs
Comment (cpu) [reply src/crl.rs]: I filed an explicit issue for this so it can be referenced with a comment and not lost to the sands of time: https://github.com/rustls/webpki/issues/80

**Pull request merged**
